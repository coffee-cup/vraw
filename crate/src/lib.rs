use wasm_bindgen::prelude::*;

mod error;
mod interpret;
mod lexer;
mod parser;
mod utils;

// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
// allocator.
#[cfg(feature = "wee_alloc")]
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

#[wasm_bindgen]
pub struct Foo {
    test: String,
}

#[wasm_bindgen]
impl Foo {
    pub fn go_riders(&self, x: &str) -> String {
        let mut s = self.test.clone();
        s.push_str(x);
        s
    }
}

#[wasm_bindgen]
pub fn bar(x: &str) -> Foo {
    let foo = Foo {
        test: "hello world".to_owned(),
    };
    foo
}

#[wasm_bindgen]
pub fn compile(input: &str) -> String {
    let tokens = match lexer::lex(&input.to_owned()) {
        Ok(tokens) => tokens,
        Err(err) => return format!("{}", err),
    };

    let program = match parser::parse_program(tokens) {
        Ok(program) => program,
        Err(err) => return format!("{}", err),
    };

    let result = match interpret::eval_program(&program) {
        Ok(value) => value,
        Err(err) => {
            return format!("{}", err);
        }
    };

    result
}
