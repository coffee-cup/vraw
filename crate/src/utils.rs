#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Pos {
    pub line: u32,
    pub column: u32,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}

pub trait HasPos {
    fn pos(&self) -> Pos;
}

pub fn create_pos(line: u32, column: u32) -> Pos {
    Pos {
        line: line,
        column: column,
    }
}

pub fn create_range(start: Pos, end: Pos) -> Range {
    Range {
        start: start,
        end: end,
    }
}

pub fn set_panic_hook() {
    // When the `console_error_panic_hook` feature is enabled, we can call the
    // `set_panic_hook` function at least once during initialization, and then
    // we will get better error messages if our code ever panics.
    //
    // For more details see
    // https://github.com/rustwasm/console_error_panic_hook#readme
    #[cfg(feature = "console_error_panic_hook")]
    console_error_panic_hook::set_once();
}
