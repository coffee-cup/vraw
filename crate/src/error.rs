use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error<T: fmt::Display> {
    pub error_type: T,
    pub pos: Pos,
}

impl<T: fmt::Display> Error<T> {
    pub fn new(error_type: T, pos: Pos) -> Error<T> {
        Error {
            error_type: error_type,
            pos: pos,
        }
    }
}

impl<T: fmt::Display> HasPos for Error<T> {
    fn pos(&self) -> Pos {
        self.pos
    }
}

impl<T: fmt::Display> fmt::Display for Error<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "[{}:{}]: {}",
            self.pos.line, self.pos.column, self.error_type
        )
    }
}
