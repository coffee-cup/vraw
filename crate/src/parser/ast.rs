use crate::utils::*;

pub type Ident = String;
pub type Arg = Ident;

#[derive(Debug, PartialEq, Clone)]
pub struct Program {
    pub decls: Vec<Decl>,
    pub end: Pos,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Decl {
    ShapeDecl(Shape),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Shape {
    pub name: Ident,
    pub args: Vec<Arg>,
    pub block: Block,
    pub pos: Pos,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Block {
    pub calls: Vec<FunCall>,
    pub range: Range,
}

#[derive(Debug, PartialEq, Clone)]
pub struct NamedArg {
    pub name: Ident,
    pub expr: Expr,
}

#[derive(Debug, PartialEq)]
pub enum Stmt {
    Expr(Expr, Pos),
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expr {
    Name(Ident, Range),
    Literal(Literal, Range),
    Binary(Box<Expr>, BinOp, Box<Expr>, Pos),
    Unary(UnOp, Box<Expr>, Pos),
    Grouping(Box<Expr>),
}

#[derive(Debug, PartialEq, Clone)]
pub struct FunCall {
    pub ident: Ident,
    pub args: Vec<NamedArg>,
    pub range: Range,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Literal {
    Number(f64),
    String(String),
}

#[derive(Debug, PartialEq, Clone)]
pub enum BinOp {
    Mul,
    Div,
    Add,
    Sub,
}

#[derive(Debug, PartialEq, Clone)]
pub enum UnOp {
    Neg,
}

pub fn prec(p: Precedence) -> u32 {
    p as u32
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Precedence {
    Conditional = 20,
    Sum = 30,
    Product = 40,
    Exponent = 50,
    Prefix = 60,
    Postfix = 70,
    Call = 80,
}

impl HasPos for Expr {
    fn pos(&self) -> Pos {
        match self {
            Expr::Name(_, range) => range.start,
            Expr::Literal(_, range) => range.start,
            Expr::Binary(_, _, _, p) => *p,
            Expr::Unary(_, _, p) => *p,
            Expr::Grouping(ref e) => e.pos(),
        }
    }
}

impl HasPos for Stmt {
    fn pos(&self) -> Pos {
        match self {
            Stmt::Expr(_, p) => *p,
        }
    }
}

impl HasPos for FunCall {
    fn pos(&self) -> Pos {
        self.range.start
    }
}

impl HasPos for Shape {
    fn pos(&self) -> Pos {
        self.pos
    }
}
