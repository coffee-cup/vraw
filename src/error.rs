use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error {
    message: String,
    pos: Option<Pos>,
}

impl Error {
    pub fn new(message: String, pos: Option<Pos>) -> Error {
        Error {
            message: message,
            pos: pos,
        }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.pos {
            Some(p) => write!(f, "[{}:{}]: {}", p.line, p.column, self.message),
            _ => write!(f, "{}", self.message),
        }
    }
}
