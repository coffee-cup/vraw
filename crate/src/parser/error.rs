use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum ParseErrorType {
    UnExpectedEndOfInput,
    IdentiferCannotBeReservedWord(String),
    UnBalancedParen,
    Expected(String, Option<String>),
}

impl fmt::Display for ParseErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            ParseErrorType::UnExpectedEndOfInput => write!(f, "Unexpected end of input."),
            ParseErrorType::IdentiferCannotBeReservedWord(id) => {
                write!(f, "Identifier {} cannot be a reserved word.", id)
            }
            ParseErrorType::UnBalancedParen => write!(f, "Unbalanced paren."),
            ParseErrorType::Expected(expected, found) => match found {
                Some(found) => write!(f, "Expected {}. Found {}.", expected, found),
                None => write!(f, "Expected {}", expected),
            },
        }
    }
}

pub type ParseError = Error<ParseErrorType>;
