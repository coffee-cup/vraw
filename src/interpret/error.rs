use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum EvalErrorType {
    VariableNotDefined(String),
    ShapeNotDefined(String),
    TypeMismatch(String, String),
    SvgExpectsString(String),
    ShapeAlreadyDefined(String),
    NumArgs(String, usize, usize),
    InvalidArgName(String, String),
    MissingArgs(String, Vec<String>),
    MissingMain,
}

impl fmt::Display for EvalErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            EvalErrorType::VariableNotDefined(name) => write!(f, "Variable `{}` not defined", name),
            EvalErrorType::ShapeNotDefined(name) => write!(f, "Shape `{}` not defined", name),
            EvalErrorType::TypeMismatch(expected, received) => write!(
                f,
                "Typemismatch. Expected: {}, Received: {}",
                expected, received
            ),
            EvalErrorType::SvgExpectsString(received) => write!(
                f,
                "svg value arg needs to be a string. Received {}",
                received
            ),
            EvalErrorType::ShapeAlreadyDefined(name) => write!(f, "Shape {} already defined", name),
            EvalErrorType::NumArgs(func, expected, received) => write!(
                f,
                "Incorrect number of args to {}. Expected: {}, Received: {}",
                func, expected, received
            ),
            EvalErrorType::InvalidArgName(func, arg) => {
                write!(f, "{} does not have an arg named {}", func, arg)
            }
            EvalErrorType::MissingArgs(func, args) => {
                write!(f, "Missing args {} for {}", args.join(", "), func)
            }
            EvalErrorType::MissingMain => write!(f, "Missing main shape"),
        }
    }
}

pub type EvalError = Error<EvalErrorType>;
