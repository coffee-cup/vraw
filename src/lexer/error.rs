use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum LexerErrorType {
    InvalidIdentifier,
    StringNeverTerminated,
    UnexpectedCharacter(char),
}

impl fmt::Display for LexerErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            LexerErrorType::InvalidIdentifier => write!(f, "Identifier must start with a letter"),
            LexerErrorType::StringNeverTerminated => write!(f, "String is never terminated"),
            LexerErrorType::UnexpectedCharacter(c) => write!(f, "Unexpected character `{}`", c),
        }
    }
}

pub type LexerError = Error<LexerErrorType>;
