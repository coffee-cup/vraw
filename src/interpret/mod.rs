use std::collections::HashMap;

use crate::parser::ast::*;
use crate::utils::*;

mod error;

use error::EvalErrorType::*;
use error::*;

#[derive(Debug, PartialEq, Clone)]
pub enum Value {
    Number(f64),
    String(String),
}

trait Name {
    fn name(&self) -> String;
}

impl Name for Value {
    fn name(&self) -> String {
        match self {
            Value::Number(_) => "number".to_owned(),
            Value::String(_) => "string".to_owned(),
        }
    }
}

type StackFrame = String;

struct Context<'a> {
    stack: Vec<StackFrame>,
    parent: Option<&'a Context<'a>>,
    scope: HashMap<String, Value>,
    shapes: HashMap<String, Shape>,
}

impl Value {
    pub fn from_number(n: f64) -> Value {
        Value::Number(n)
    }

    pub fn from_string(s: &str) -> Value {
        Value::String(s.to_owned())
    }
}

type EvalResult<T> = Result<T, EvalError>;

fn eval_error<T>(error_type: EvalErrorType, pos: Pos) -> EvalResult<T> {
    Err(EvalError::new(error_type, pos))
}

impl<'a> Context<'a> {
    pub fn new() -> Self {
        Context {
            stack: vec![],
            parent: None,
            scope: HashMap::new(),
            shapes: HashMap::new(),
        }
    }

    pub fn with_parent(ctx: &'a Context) -> Self {
        Context {
            stack: ctx.stack.clone(),
            parent: Some(ctx),
            scope: HashMap::new(),
            shapes: HashMap::new(),
        }
    }

    pub fn get(&self, key: &str) -> Option<Value> {
        if let Some(x) = self.scope.get(key) {
            Some(x.clone())
        } else if let Some(p) = self.parent {
            p.get(key)
        } else {
            None
        }
    }

    pub fn set(&mut self, key: &str, value: Value) {
        self.scope.insert(key.to_owned(), value);
    }
}

fn eval_literal(lit: &Literal) -> EvalResult<Value> {
    match lit {
        Literal::Number(n) => Ok(Value::Number(*n)),
        Literal::String(s) => Ok(Value::from_string(s)),
    }
}

fn get_number(value: Value, pos: Pos) -> EvalResult<f64> {
    match value {
        Value::Number(n) => Ok(n),
        v => return eval_error(TypeMismatch("number".to_owned(), v.name()), pos),
    }
}

fn get_string(value: Value, pos: Pos) -> EvalResult<String> {
    match value {
        Value::String(s) => Ok(s),
        v => return eval_error(TypeMismatch("string".to_owned(), v.name()), pos),
    }
}

fn eval_binary(
    op: BinOp,
    lhs_expr: &Expr,
    rhs_expr: &Expr,
    ctx: &mut Context,
) -> EvalResult<Value> {
    let lhs = eval_expression(lhs_expr, ctx)?;
    let rhs = eval_expression(rhs_expr, ctx)?;

    let lhs = get_number(lhs, lhs_expr.pos())?;
    let rhs = get_number(rhs, rhs_expr.pos())?;

    match op {
        BinOp::Mul => Ok(Value::Number(lhs * rhs)),
        BinOp::Div => Ok(Value::Number(lhs / rhs)),
        BinOp::Add => Ok(Value::Number(lhs + rhs)),
        BinOp::Sub => Ok(Value::Number(lhs - rhs)),
    }
}

fn eval_unary(op: UnOp, expr: &Expr, ctx: &mut Context) -> EvalResult<Value> {
    let value = eval_expression(expr, ctx)?;
    let value = get_number(value, expr.pos())?;

    match op {
        UnOp::Neg => Ok(Value::from_number(value * -1.0)),
    }
}

fn eval_expression(expr: &Expr, ctx: &mut Context) -> EvalResult<Value> {
    match expr {
        Expr::Name(n, r) => match ctx.get(n) {
            Some(value) => Ok(value),
            None => return eval_error(VariableNotDefined(n.clone()), r.start),
        },
        Expr::Literal(lit, _) => eval_literal(lit),
        Expr::Binary(ref lhs, op, ref rhs, _) => eval_binary(op.clone(), lhs, rhs, ctx),
        Expr::Unary(op, ref expr, _) => eval_unary(op.clone(), expr, ctx),
        Expr::Grouping(ref expr) => eval_expression(expr, ctx),
    }
}

fn eval_svg_call(call: &FunCall, ctx: &mut Context) -> EvalResult<Value> {
    if call.args.len() != 1 {
        return eval_error(NumArgs(call.ident.clone(), 1, call.args.len()), call.pos());
    }

    let arg = &call.args[0];
    if arg.name != "value" {
        return eval_error(
            MissingArgs(call.ident.clone(), vec!["value".to_owned()]),
            call.pos(),
        );
    }

    let value = eval_expression(&arg.expr, ctx)?;

    let value = match get_string(value, arg.expr.pos()) {
        Err(e) => match &e.error_type {
            TypeMismatch(_, received) => {
                return eval_error(SvgExpectsString(received.clone()), e.pos())
            }
            error_type => return eval_error(error_type.clone(), e.pos()),
        },
        Ok(v) => v,
    };

    Ok(Value::String(value))
}

fn eval_call(call: &FunCall, ctx: &mut Context) -> EvalResult<Value> {
    if call.ident == "svg" {
        eval_svg_call(call, ctx)
    } else {
        match ctx.shapes.get(&call.ident) {
            _ => eval_error(ShapeNotDefined(call.ident.clone()), call.pos()),
        }
    }
}

fn eval_block(block: &Block, ctx: &mut Context) -> EvalResult<Value> {
    println!("TEST");
    let mut out: String = "".to_owned();

    for call in block.calls.iter() {
        match eval_call(call, ctx)? {
            Value::String(s) => out.push_str(s.as_str()),
            _ => panic!("call should return string value."),
        }
    }

    Ok(Value::String(out))
}

fn find_shapes(program: &Program) -> EvalResult<HashMap<String, Shape>> {
    let mut shapes: HashMap<String, Shape> = HashMap::new();
    let mut found_main = false;

    for decl in program.decls.iter() {
        match &decl {
            Decl::ShapeDecl(shape) => {
                if shapes.contains_key(&shape.name) {
                    return eval_error(ShapeAlreadyDefined(shape.name.clone()), shape.pos());
                }

                if shape.name == "main" {
                    found_main = true;
                }

                shapes.insert(shape.name.clone(), shape.clone());
            }
        };
    }

    if !found_main {
        return eval_error(MissingMain, program.end);
    }

    Ok(shapes)
}

pub fn eval_program(program: &Program) -> EvalResult<Value> {
    let ctx = &mut Context::new();

    let shapes = find_shapes(program)?;
    ctx.shapes = shapes;

    let main = ctx.shapes.get("main").unwrap().clone();
    let value = eval_block(&main.block, ctx)?;

    Ok(value)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer;
    use crate::parser;

    fn check_expression(line: &str, expected: Value) {
        let tokens = lexer::lex(&line.to_owned()).unwrap();
        let expr = parser::parse_expression(tokens).unwrap();
        let output = eval_expression(&expr, &mut Context::new()).unwrap();

        assert_eq!(output, expected);
    }

    #[test]
    fn context_empty() {
        let context = Context::new();
        assert_eq!(context.scope.len(), 0);
    }

    #[test]
    fn adding_item_to_context() {
        let mut context = Context::new();
        context.set("foo", Value::from_string("bar"));
        assert_eq!(context.get("foo"), Some(Value::from_string("bar")));
    }

    #[test]
    fn literals() {
        check_expression("1", Value::Number(1.0));
    }

    #[test]
    fn binary_ops() {
        check_expression("1 + 1", Value::Number(2.0));
        check_expression("1.5 + 1", Value::Number(2.5));
        check_expression("2 * 2", Value::Number(4.0));
        check_expression("4 / 2", Value::Number(2.0));
        check_expression("6 - 3", Value::Number(3.0));
        check_expression("3 * (2 + -4) / 2 * 3", Value::Number(-9.0));
    }

    #[test]
    fn simple_find_shapes() {
        let line = "
shape main() {}
shape circle(r) {}
shape rect(w, h) {}
";
        let tokens = lexer::lex(&line.to_owned()).unwrap();
        let program = parser::parse_program(tokens).unwrap();

        let shapes = find_shapes(&program).unwrap();

        assert_eq!(shapes.len(), 3)
    }

    #[test]
    fn shape_already_defined() {
        let line = "
shape main() {}
shape circle(r) {}
shape circle(r) {}
";
        let tokens = lexer::lex(&line.to_owned()).unwrap();
        let program = parser::parse_program(tokens).unwrap();

        match find_shapes(&program) {
            Ok(_) => assert!(false),
            Err(_) => assert!(true),
        };
    }

    #[test]
    fn missing_main_shape() {
        let line = "
shape circle(r) {}
";
        let tokens = lexer::lex(&line.to_owned()).unwrap();
        let program = parser::parse_program(tokens).unwrap();

        match find_shapes(&program) {
            Ok(_) => assert!(false),
            Err(e) => match e.error_type {
                MissingMain => assert!(true),
                _ => assert!(false),
            },
        };
    }

    #[test]
    fn eval_simple_program() {
        let line = "
shape main() {
  svg(value: 3)
}
";

        let tokens = lexer::lex(&line.to_owned()).unwrap();
        let program = parser::parse_program(tokens).unwrap();

        let ctx = &mut Context::new();
        let value = eval_program(&program).unwrap();

        assert_eq!(value, Value::String("tes".to_owned()));
    }
}
