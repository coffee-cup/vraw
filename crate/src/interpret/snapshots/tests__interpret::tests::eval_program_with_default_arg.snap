---
created: "2019-07-12T18:25:23.583239Z"
creator: insta@0.8.1
source: src/interpret/mod.rs
expression: value
---
"<svg width=\"100%\" height=\"100%\" xmlns=\"http://www.w3.org/2000/svg\">hello</svg>"
